src/jit/namedintrinsiclist.h

```diff
@@ -16,10 +16,70 @@ enum NamedIntrinsic : unsigned int
     NI_System_Collections_Generic_EqualityComparer_get_Default = 4,
 #if FEATURE_HW_INTRINSICS
     NI_HW_INTRINSIC_START,
-#define HARDWARE_INTRINSIC(id, name, isa) NI_##id,
+#define HARDWARE_INTRINSIC(id, name, isa, ret, num, ival, size, t1, t2, t3, t4, t5, t6, t7, t8, t9, t10, cate) NI_##id,
 #include "hwintrinsiclistxarch.h"
     NI_HW_INTRINSIC_END
 #endif
 };
 
+#if FEATURE_HW_INTRINSICS && defined(_TARGET_XARCH_)
+enum HWIntrinsicCategory : unsigned int
+{
+    // Simple SIMD intrinsics
+    // - take Vector128/256<T> parameters
+    // - return a value of Vector128/256<T>
+    // - all the vector parameters and return value has same vector size
+    // - generate single instruction
+    // - the codegen of overloads can be determined by intrinsicID and base type of returned vector
+    // - not any other kind intrinsics
+    HW_Category_SimpleSIMD = 0,
```

Comment:
From my own investigations, some of these generalizations don't work well for all instructions and can often make containment analysis harder.

For example, this doesn't cover things like: `am I commutative` (which indicates we can swap op1 and op2 if op1 is containable, but op2 isn't).

It also doesn't cover differences like `MoveHighToLow`, which qualifies as a `simple intrinsic`, but does not allow a memory encoding.

Reply:
Yes, actually I am considering to give simple intrinsic a mask bit rather than 0.

Reply:
Maybe we should have a `byte category`, `byte encoding`, and a `ushort flags` field.

category is, as you have it.
encoding helps identify things like whether the instruction supports `Reg, Reg` or `Reg, Reg/Mem`
flags helps identify things like whether the instruction is commutative

Reply:
They are all used for table-driven, just giving mask bits looks enough now.
And 8-bit may be not enough for the category field.
Why do we need the “flags”? “commutative” can be a category bit.

Reply:
It helps reduce code duplication and the number of checks required when they are distinct fields

For example, when doing containment analysis, the most important information is which operands can be contained. With mask bits, we will need to have multiple comparison checks against the various categories (`if ((category == SIMDScalar_SecondOpContainable) || (category == Scalar_SecondOpContainable))`

The containment analysis, I think, could look like this (pseudo-code):
```
switch (encoding)
{
    case R_R_RM:  // target must be reg, first operand must be reg, second op can be reg or mem
        if (!op2->isContainable() && (flags & commutative))
        {
            swapOperands();
        }

        // TODO: SSE/SSE2 instructions on non-vex are technically R_RM_RM when `targetReg != op1Reg` since we need an additional `movaps` or `movups` instruction

        if (op2->isContainable() && (isVexEncodingSupported() || (category != SIMDVector))) // on non-vex, the memory operand for SIMDVector must be aligned, ideally we would detect the case where it is aligned and allow containment
        {
            makeContained(op2);
        }
        break;

    // other encodings
}
```


Once containment analysis has been done, this ends up being important information for codegen as well.
 
The codegen can then look something like (also pseudo-code):
```
switch (encoding)
{
    case R_R_RM:
        genHWIntrinsic_R_R_RM(node);
        break;

    // other encodings
}

genHWIntrinsic_R_R_RM(node)
{
    if (op2->isContained())
    {
        // NOTE: most of the "logic" is fairly consistent between things like `R_R_RM` and `R_R_RM_I`, the only difference is whether we call `emit->emitIns_R_R_*` or `emit->emitIns_R_R_*_I`
        //       we should take advantage of this

        // logic
            emit->emitIns_R_R_C();
        // logic
            emit->emitIns_R_R_S();
        // logic
            emit->emitIns_R_R_A();
    }
    else
    {
        emit->emitIns_R_R_R();
    }
}
```

Reply:
Basically, I've found the most important piece of information to be the final encodings supported (this is the primary factor in the control flow for our logic).

We then make slight tweaks to that based on the category (am I SIMD, SIMDScalar, Scalar, etc) or flags (am I commutative, do I require alignment, etc) set.